File: code/Java/command/control/RemoteControl.java
package command.control;

import command.command.Command;
import command.command.NoCommand;

public class RemoteControl {
    private Command[] onCommands;
    private Command[] offCommands;
    private Command undoCommand;

    public RemoteControl(int n) {
        onCommands = new Command[n];
        offCommands = new Command[n];

        Command defaultCommand = new NoCommand();

        for(int i=0; i<n; i++) {
            onCommands[i] = defaultCommand;
            offCommands[i] = defaultCommand;
        }
        undoCommand = defaultCommand;
    }

    public void setCommand(int i, Command on, Command off) {
        if(i >= onCommands.length || i < 0)  return;
        onCommands[i] = on;
        offCommands[i] = off;
    }

    public void onButtonWasPushed(int i) {
        if(i >= onCommands.length || i < 0)  return;
        onCommands[i].execute();
        undoCommand = onCommands[i];
    }

    public void offButtonWasPushed(int i) {
        if(i >= onCommands.length || i < 0)  return;
        offCommands[i].execute();
        undoCommand = offCommands[i];
    }

    public void undoButtonWasPushed() {
        undoCommand.undo();     // !!!
    }

    public String toString() {
        StringBuffer stringBuff = new StringBuffer();
        stringBuff.append("\n------ Remote Control -------\n");
        for (int i = 0; i < onCommands.length; i++) {
            stringBuff.append("[slot " + i + "] " + onCommands[i].getClass().getName()
                    + "    " + offCommands[i].getClass().getName() + "\n");
        }
        stringBuff.append("[undo] " + undoCommand.getClass().getName() + "\n");
        return stringBuff.toString();
    }
}

File: code/C++/command/remotecontrol.cpp
#include "remotecontrol.h"

RemoteControl::RemoteControl(int n) {
	defaultCommand = new NoCommand();

	onCommands = std::vector<Command*>(n, defaultCommand);
	offCommands = std::vector<Command*>(n, defaultCommand);
	undoCommand = defaultCommand;
}

RemoteControl::~RemoteControl()
{
	if (defaultCommand) {
		delete defaultCommand;
		defaultCommand = nullptr;
	}
}

void RemoteControl::setCommand(int i, Command* const on, Command* const off) {
	if (i < 0 || i >= onCommands.size())	return;
	onCommands[i] = on;
	offCommands[i] = off;
}

void RemoteControl::onButtonWasPushed(int i) {
	if (i < 0 || i >= onCommands.size())	return;
	onCommands[i]->execute();
	undoCommand = onCommands[i];
}

void RemoteControl::offButtonWasPushed(int i) {
	if (i < 0 || i >= onCommands.size())	return;
	offCommands[i]->execute();
	undoCommand = offCommands[i];
}

void RemoteControl::undoButtonWasPushed() {
	undoCommand->undo();
}

void RemoteControl::display()
{
	std::cout << "\n------ Remote Control -------\n" << std::endl;
	for (int i = 0; i < onCommands.size(); i++) {
		std::cout << "[slot " << i << "] " << typeid(*dynamic_cast<Command*>(onCommands[i])).name() << 
			"    " << typeid(*dynamic_cast<Command*>(offCommands[i])).name() << std::endl;
	}
	std::cout << "[undo] " << typeid(*dynamic_cast<Command*>(undoCommand)).name() << std::endl;
	//std::cout << "[undo] " << typeid(undoCommand).name() << std::endl;
}